can-bus/common.c

```diff
@@ -0,0 +1,217 @@
+    va_end(args);
+}
+
+/* Timing function required by ISO-TP to compile */
+uint32_t isotp_user_get_ms(void) {
+    struct timeval t;
+    gettimeofday(&t, NULL);
+    return (t.tv_sec * 1000) + (t.tv_usec / 1000);
+}
+
+/* CAN bus send function require by ISO-TP to compile */
+int isotp_user_send_can(const uint32_t arbitration_id, const uint8_t* data,
+        const uint8_t size) {
+    /* Copy the message into a CAN bus frame, the message will always be
+     * 8 bytes or less */
+    struct can_frame frame;
+    frame.can_id = arbitration_id;
+    frame.can_dlc = size;
+    memcpy(frame.data, data, size);
+
+    /* Write the frame to the CAN bus */
+    if (write(sock, &frame, sizeof(struct can_frame)) != sizeof(struct can_frame)) {
+       perror("Write error\n");
+       return -1;
+    }
+
+    return 0;
+}
+
+/* Our CAN bus receive function */
+int can_receive(uint8_t data[8], int *length) {
+    int nbytes;
+    struct can_frame frame;
+    struct pollfd p[1];
+
+	p[0].fd = sock;
+	p[0].events = POLLIN;
+
+    /* Poll for new data */
+	int retval = poll(p, 1, 10);
+
+    if (retval < 0) {
+        perror("Poll error\n");
+        return 1;
+    }
+    else if (retval == 0) {
+        // No data
+        *length = 0;
+        return EAGAIN;
+    }
+
+    /* Read in the frame data */
+    nbytes = read(sock, &frame, sizeof(struct can_frame));
+    if (nbytes < 0) {
+       perror("Read error\n");
+       return 1;
+    }
+    memcpy(data, frame.data, frame.can_dlc);
+    *length = frame.can_dlc;
+    return 0;
+}
+
+/* Connect to the CAN bus */
+int can_connect(const char *address, uint16_t filter)
+{
+    struct can_filter rfilter[1];
+    rfilter[0].can_id   = filter;
+    rfilter[0].can_mask = 0xFFF;
+
+    if ((sock = socket(PF_CAN, SOCK_RAW, CAN_RAW)) < 0) {
+       perror("Socket open error\n");
+       return 1;
+    }
+    setsockopt(sock, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter));
+    struct ifreq ifr;
+
+    strcpy(ifr.ifr_name, address);
+    ioctl(sock, SIOCGIFINDEX, &ifr);
+
+    struct sockaddr_can addr;
+    memset(&addr, 0, sizeof(addr));
+    addr.can_family = AF_CAN;
+    addr.can_ifindex = ifr.ifr_ifindex;
+    if (bind(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+       perror("Bind error\n");
+       return 1;
+    }
+
+    /* Setup the receive copy buffer */
+    copy_buf = malloc(REC_BUF_SIZE);
+    copy_buf_ptr = copy_buf;
+    copy_buf_len = 0;
+    if (!copy_buf) {
+        printf("Copy buf malloc fail\n");
+        return -99;
+    }
+
+    return 0;
+}
+
+void can_close()
+{
+    free(copy_buf);
+    close(sock);
+}
+
+/* The wolfSSL send packet callback. This function sends the data via ISO-TP
+ * and then listens and processes the ISO-TP control packets until the message
+ * is completely sent */
+int send_ssl(WOLFSSL *ssl, char *buf, int sz, void *ctx)
+{
+    uint8_t data[8];
+    int length;
+    IsoTpLink *g_link = (struct IsoTpLink*)ctx;
+    int ret = isotp_send(g_link, buf, sz);
```

Comment:
`buf` herę needs to be unsigned. Consider casting to `uint8_t`. Add `-Wall -Wextra` to your Makefile to reproduce.